### ISSUE TITLE
I must be doing something wrong...

### ISSUE THREAD

**Issue by suddenelfilio** [structuremap/structuremap#565]
Okay I have

an interface IDomainEvents
```
 public interface IDomainEvent
    {
    }

```
an interface IDomainEventHandler

```
 public interface IDomainEventHandler<T> where T: IDomainEvent
    {
        void Handle(T args);
    }

```
an event class
```
  public class DataChangedEvent : IDomainEvent
    {
        public int Id { get; set; }

    }
```

and my program

````
class Program
    {
        private static IContainer _container;

        static void Main(string[] args)
        {
            _container = ConfigureContainer();

            Console.WriteLine(_container.GetInstance<IDomainEventHandler<IDomainEvent>>() as DefaultEventsHandler<IDomainEvent> != null);
            Console.WriteLine(_container.GetInstance<IDomainEventHandler<DataChangedEvent>>() as DataEventHandler != null);
            Console.ReadLine();

        }

        private static IContainer ConfigureContainer()
        {
            return new Container(x =>
            {
                x.Scan(scanner =>
                {
                    scanner.TheCallingAssembly();
                    scanner.WithDefaultConventions();
                    scanner.ConnectImplementationsToTypesClosing(typeof(IDomainEventHandler<>));
                });
                x.For(typeof(IDomainEventHandler<>)).Use(typeof(DefaultEventsHandler<>));
            });
        }


        public class DefaultEventsHandler<T> : IDomainEventHandler<T> where T : IDomainEvent
        {
            public virtual void Handle(T args)
            {
                Console.ForegroundColor = ConsoleColor.Blue;
                Console.WriteLine(args.GetType().Name);
                Console.ResetColor();
            }
        }

        public class DataEventHandler : IDomainEventHandler<DataChangedEvent>
        {
            public void Handle(DataChangedEvent args)
            {
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine(args.Id);
                Console.ResetColor();
            }
        }
    }
````

If i run this I would expect:

True
True

except I'm getting

True
False

so the closed type is not picked up here? Did I misunderstand this functionality?

**Comment by suddenelfilio** [comment]
Okay I've update the sample to include the test in the bugs https://github.com/structuremap/structuremap/blob/master/src/StructureMap.Testing/Bugs/ConnectImplementationsToTypesClosing_is_wonky_in_Registry_added_by_Configure.cs for issue #525 

And it behaves the same

```
   class Program
    {
        private static IContainer _container;

        static void Main(string[] args)
        {
            _container = ConfigureContainer();

            Console.WriteLine(_container.GetInstance<IDomainEventHandler<IDomainEvent>>() as DefaultEventsHandler<IDomainEvent> != null);
            Console.WriteLine(_container.GetInstance<IDomainEventHandler<DataChangedEvent>>() as DataEventHandler != null);

            Console.WriteLine("");

            Console.WriteLine(_container.GetInstance<IAmOpenGeneric<int>>() as TheClosedGeneric<int> != null);
            Console.WriteLine(_container.GetInstance<IAmOpenGeneric<string>>() as SpecificClosedGeneric != null);
            Console.ReadLine();

        }

        private static IContainer ConfigureContainer()
        {
            return new Container(x =>
            {
                x.Scan(scanner =>
                {
                    scanner.TheCallingAssembly();
                    scanner.WithDefaultConventions();
                    scanner.ConnectImplementationsToTypesClosing(typeof(IDomainEventHandler<>));
                    scanner.ConnectImplementationsToTypesClosing(typeof(IAmOpenGeneric<>));
                });
                x.For(typeof(IDomainEventHandler<>)).Use(typeof(DefaultEventsHandler<>));
                x.For(typeof(IAmOpenGeneric<>)).Use(typeof(TheClosedGeneric<>));
            });
        }



        public interface IAmOpenGeneric<T>
        {
        }

        public class TheClosedGeneric<T> : IAmOpenGeneric<T>
        {
        }

        public class SpecificClosedGeneric : TheClosedGeneric<string>
        {
        }


        public class DefaultEventsHandler<T> : IDomainEventHandler<T> where T : IDomainEvent
        {
            public virtual void Handle(T args)
            {
                Console.ForegroundColor = ConsoleColor.Blue;
                Console.WriteLine(args.GetType().Name);
                Console.ResetColor();
            }
        }

        public class DataEventHandler : DefaultEventsHandler<DataChangedEvent>
        {
            public override void Handle(DataChangedEvent args)
            {
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine(args.Id);
                Console.ResetColor();
            }
        }
    }
```

![image](https://cloud.githubusercontent.com/assets/6460448/25037711/f31d3772-20fa-11e7-8bb0-7e4dea942209.png)

**Comment by jeremydmiller** [comment]
Okay, so a couple things:

* Can you attach a failing unit test instead of the console app? That's just too hard to parse and the first thing I'd do anyway is to switch to a test
* Can you write out the WhatDoIHave() output for the container and see if that tells you anything first?

**Comment by suddenelfilio** [comment]
Okay so this gets stranger by the minute

I wrote up a test

 ```
  public class TheTruth
    {
        [Fact]
        public void AndNothingButTheTruth()
        {
            var container = new Container(x =>
             {
                 x.Scan(scanner =>
                 {
                     scanner.TheCallingAssembly();
                     scanner.WithDefaultConventions();
                     scanner.ConnectImplementationsToTypesClosing(typeof(Program.IAmOpenGeneric<>));
                 });
                 x.For(typeof(Program.IAmOpenGeneric<>)).Use(typeof(Program.TheClosedGeneric<>));
             });

            container.GetInstance<Program.IAmOpenGeneric<int>>().ShouldBeOfType<Program.TheClosedGeneric<int>>();
            container.GetInstance<Program.IAmOpenGeneric<string>>().ShouldBeOfType<Program.SpecificClosedGeneric>();
        }
    }

```
and this fails
![image](https://cloud.githubusercontent.com/assets/6460448/25044319/59ef6634-2126-11e7-9d60-4567600c5fb5.png)

now if I move the respective interface and classes into the test file it passes

```
 public class TheTruth
    {
        [Fact]
        public void AndNothingButTheTruth()
        {
            var container = new Container(x =>
             {
                 x.Scan(scanner =>
                 {
                     scanner.TheCallingAssembly();
                     scanner.WithDefaultConventions();
                     scanner.ConnectImplementationsToTypesClosing(typeof(IAmOpenGeneric<>));
                 });
                 x.For(typeof(IAmOpenGeneric<>)).Use(typeof(TheClosedGeneric<>));
             });

            container.GetInstance<IAmOpenGeneric<int>>().ShouldBeOfType<TheClosedGeneric<int>>();
            container.GetInstance<IAmOpenGeneric<string>>().ShouldBeOfType<SpecificClosedGeneric>();
        }
    }

    public interface IAmOpenGeneric<T>
    {
    }

    public class TheClosedGeneric<T> : IAmOpenGeneric<T>
    {
    }

    public class SpecificClosedGeneric : TheClosedGeneric<string>
    {
    }
```
![image](https://cloud.githubusercontent.com/assets/6460448/25044366/9a381ab0-2126-11e7-875e-aad5e02b5027.png)

**Comment by suddenelfilio** [comment]
OMG!!! I found my culprit. The interface and classes where in the Program class making them Program.IAmOpenGeneric<T> instead of IAmOpenGeneric. If I move them out of the Program class it all works fine...

So the answer is yes I was missing something.

**Comment by jeremydmiller** [comment]
@suddenelfilio Cool, mind if I close this one?

**Comment by suddenelfilio** [comment]
well actually I am surprised it can resolve to Program.IAmOpenGeneric, but not the closed implementation Program.SpecificClosedGeneric type. Should this  not work as expected regardsless of nesting in classes?

**Comment by jeremydmiller** [comment]
Being or not being an inner class has no impact other than if you're running this from SM's codebase itself, it keeps you from colliding on type names from other tests
